CounterStrike/src/dev/danablend/counterstrike/listeners/EntityPickupItemListener.java

```diff
@@ -34,53 +34,81 @@ public void onPlayerPickupItem(EntityPickupItemEvent e) {
+                e.getItem().remove();
+
+                if (player.getInventory().getItem(4) == null) {
+                    item.setAmount(1);
+                    player.getInventory().setItem(4, item);
+                }
+                e.setCancelled(true); //to cancel event, if has bomb doesn't get another, if has none already picked it up
                 return;
+
             } else if (cp.getTeam().equals(TeamEnum.COUNTER_TERRORISTS) && item.getType().equals(Material.TNT)) {
                 e.setCancelled(true);
+                return;
             }
 
             if (Weapon.isWeapon(item)) {
                 Weapon weapon = Weapon.getByItem(item);
 
                 switch (weapon.getWeaponType()) {
                     case RIFLE:
-                        if (p.getInventory().getItem(0) == null) {
+                        if (player.getInventory().getItem(0) == null) {
                             e.getItem().remove();
-                            p.getInventory().setItem(0, item);
+                            player.getInventory().setItem(0, item);
                         }
-                        e.setCancelled(true);
+                        e.setCancelled(true); //to cancel event, if has Rifle doesn't get another, if has none already picked it up
                         break;
 
                     case PISTOL:
-                        if (p.getInventory().getItem(1) == null) {
+                        if (player.getInventory().getItem(1) == null) {
                             e.getItem().remove();
-                            p.getInventory().setItem(1, item);
+                            player.getInventory().setItem(1, item);
                         }
-                        e.setCancelled(true);
+                        e.setCancelled(true); //to cancel event, if has PISTOL doesn't get another, if has none already picked it up
                         break;
 
                     case GRENADE:
+
+                        if (player.getInventory().getItem(3) == null) {
+                            e.getItem().remove();
+                            player.getInventory().setItem(3, item);
+                        } else {
+                            ItemStack items = player.getInventory().getItem(3);
+
+                            if (items.getItemMeta().getDisplayName().equals(item.getItemMeta().getDisplayName())) {
+                                e.getItem().remove();
+                                items.setAmount(item.getAmount() + item.getAmount());
+                                player.getInventory().setItem(3, items);
+                            } else {
+                               // player.sendMessage("Unmatch grenade. Discarding...");
+                            }
+                        }
+                        e.setCancelled(true);
                         break;
 
                     default:
                         e.setCancelled(true);
                         break;
                 }

```

Comment:
`e.setCancelled` can be extracted and moved after the "switch", as it is executed in every case.